src/dune_rules/dialect.ml

```diff
@@ -78,13 +81,24 @@ let decode =
       field_o
         "format"
         (map ~f:(fun (loc, x) -> loc, x, []) (located Action.decode_dune_file))
+    and+ merlin_reader = field_o "merlin_reader" (located (repeat1 string))
     and+ syntax_ver = Syntax.get_exn Stanza.syntax in
-    let ver = 3, 9 in
-    if syntax_ver < ver && Option.is_some (String.index_from extension 1 '.')
-    then (
-      let what = "the possibility of defining extensions containing periods" in
-      Syntax.Error.since loc Stanza.syntax ver ~what);
-    { File_kind.kind; extension; preprocess; format }
+    let () =
+      let ver = 3, 9 in
+      if syntax_ver < ver && Option.is_some (String.index_from extension 1 '.')
+      then (
+        let what = "the possibility of defining extensions containing periods" in
+        Syntax.Error.since loc Stanza.syntax ver ~what)
+    in
+    let () =
+      let ver = 3, 10 in
+      match syntax_ver < ver, merlin_reader with
+      | true, Some (loc, _) ->
+        let what = "the (merlin_reader) stanza" in
+        Syntax.Error.since loc Stanza.syntax ver ~what
+      | _ -> ()
+    in
```

Comment:
The error handling check that you imitated here is specially tricky; in your case you can do something simpler, such as in https://github.com/ocaml/dune/blob/0219fad9ae3d7e45e5cd37616bfb126b9c96d802/src/dune_rules/dune_file.ml#L76-L80